tensorrt_llm/_torch/modules/attention.py

```diff
@@ -208,36 +221,25 @@ def forward(
             if qkv_lora is not None:
                 qkv = qkv + qkv_lora
 
-        if is_fused_qkv:
-            if self.pos_embd_params is None and position_ids is not None:
-                q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size],
-                                    dim=-1)
-                q, k = self.rotary_emb(position_ids, [q, k], attn_metadata)
-                qkv = torch.concat([q, k, v], dim=-1)
-
-            out_scale = None
-            if self.o_proj.has_fp8_qdq or self.o_proj.has_nv_fp4 or self.o_proj.has_fp8_block_scales:
-                out_scale = self.o_proj.inv_input_scale
-            attn_output = self.attn.forward(qkv,
-                                            None,
-                                            None,
-                                            attn_metadata,
-                                            out_scale=out_scale,
-                                            attention_mask=attention_mask,
-                                            mrope_config=mrope_config)
-        else:
+        q, k, v = qkv, None, None
+
+        if self.rotary_emb is not None and position_ids is not None:
             q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size],
                                 dim=-1)
+            q, k = self.rotary_emb(position_ids, [q, k])
 
-            if self.pos_embd_params is None and position_ids is not None:
-                q, k = self.rotary_emb(position_ids, [q, k], attn_metadata)
+        out_scale = None
+        if self.o_proj.has_fp8_qdq or self.o_proj.has_nv_fp4 or self.o_proj.has_fp8_block_scales:
+            out_scale = self.o_proj.inv_input_scale
 
-            attn_output = self.attn.forward(q.contiguous(),
-                                            k.contiguous(),
-                                            v.contiguous(),
-                                            attn_metadata,
-                                            attention_mask=attention_mask,
-                                            mrope_config=mrope_config)
+        q, k, v = self.convert_qkv(q, k, v)
+        attn_output = self.attn.forward(q,
```

Comment:
@yuxianq  
Another question: do we need to include `.contiguous()` after q, k, v when passing to `attn.forward`  after applying rope

cc. @hlu1

Reply:
It depends on if the attention kernel requires contiguous inputs or strided inputs. A common practice is for the attention kernel to call `.contiguous()` inside if contiguous inputs are required.

Reply:
I see.. do you know whether flashinfer and trtllm attn backend support strided input? AFAIK, there seems no `contiguous()` call inside both attn kernels? (I may miss something)